keps/sig-storage/0661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,808 @@
+- Resizing a PVC associated with a statefulset directly by modifying the PVC spec will not affect the 
+  Statefulset `volumeClaimTemplates`. 
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Before we proceed to the proposal it is imperative to layout the various components involved
+in creating and managing the persistent volumes associated with `StatefulSet`. 
+
+- StatefulSet controller creates the PVCs from `volumeClaimTemplates` and associates them with the pods
+spawned from it. Reconciliation logic runs for every change made to the StatefulSet
+`Spec` and `Status` reflects the progress towards making the real world state equal to the
+`Spec`.
+- The admission controller in the API server is responsible for restricting the fields
+which can be edited in a StatefulSet specification.
+- The RBAC specification in the controller policy is responsible for adding the required 
+RBAC permission for StatefulSet.
+- The StatefulSet controller also supports revision control.
+
+The following changes are proposed to achieve the goal:
+1. Make the `volumeClaimTemplates` storage size editable by modification to the api server
+   validation. Some validation checks will depend on the design decisions made in [KEP 1790](../1790-recover-resize-failure/README.md) to prevent user from shrinking the volume.
+2. Add reconciliation of the associated PVC's size and the individual `volumeClaimTemplates` size into the 
+   StatefulSet controller reconciliation.
+3. Add PVC `update` privilege to the StatefulSet controller's clusterrole RBAC.
+
+Once the above changes are made, modification to any of the `volumeClaimTemplates` storage
+size by user will update the underlying PVC via the reconciliation performed by (2) above.
+The existing PVC resize workflow kicks in at this point and attempts to resize the volume.
+
+### Error indications to user
+
+Whether a resize works or not would depend on many factors such as is `allowExpansion` set in the
+storage class, does the underlying storage allow expansion etc. Such failures will be reported in
+the corresponding PVC events. StatefulSet would only report any failures in the update operation
+performed to the PVC object.
+
+### Revision control
+
+StatefulSet supports [revision control](https://kubernetes.io/blog/2017/09/kubernetes-statefulsets-daemonsets/).
+The storage size change will not be added to the fields supporting revision control. Any change in
+revision - going forward to a new version or going back to an older version would not make any
+changes to the persistent volume size. Any changes requiring resize of the volume will need editing
```

Comment:
Can you describe why you don't want to add that to revision control?

[My assumption is that adding it would effectively block rollback, as it would mean shrinking PVC. If that's the case, please add to the KEP. If not, please add your thinking behind that.]

Reply:
What happens if user requests larger volume size and may be changes version of image in same edit? Are these operations going to be serialized one-after-other or applied at once for each replica of statefulset?